migrations: add migration to remove pre-19.2 FK representation

ajwerner opened the pull request:
We've been carrying this old representation for a long time. This migration
is written as simply as I can think to do it. It could perhaps be better.

One thing this patch doesn't do is really enforce that it's gone. Probably
we'll want to do that in validation for this cycle. In the next release we'll
stop decoding the fields.

The change should not be user visible so no release note.

Release note: None

Commit:
migrations: add migration to remove pre-19.2 FK representation

We've been carrying this old representation for a long time. This migration
is written as simply as I can think to do it. It could perhaps be better.

One thing this patch doesn't do is really enforce that it's gone. Probably
we'll want to do that in validation for this cycle. In the next release we'll
stop decoding the fields.

The change should not be user visible so no release note.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60510)
<!-- Reviewable:end -->

thoszhang reviewed (commented):
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60510)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MTImcTLDskL0WuhKTEj:-MTImcTLDskL0WuhKTEk:b-26fh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ce7be2a8ef42a60f91045cf1e9985785940b7f/pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go#L99)):*
> ```Go
>    AND EXISTS(SELECT * FROM upserted_db_namespace_entry)
>    AND EXISTS(SELECT * FROM upserted_table_namespace_entries);`,
> 		pq.Array(strings.Split(descriptorsToInject, "\n")))
> ```

I have to say I am not a fan of the huge queries, especially since they actually mutate state. I think we'll want a test that's easier to understand and modify if we run into problems with the migration. I'm fine with using the builtins but in a sequence of smaller queries, which will be a pain but will make for a better test in the long run, I think.

Arguably there should be a testing library that can do all the work of injecting descriptors/namespace entries into a test server.
___
*[pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go, line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MTIkNOuCKST17cMAZvJ:-MTIkNOuCKST17cMAZvK:bua9poa) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ce7be2a8ef42a60f91045cf1e9985785940b7f/pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go#L141)):*
> ```Go
>     CREATE TABLE parent (i INT8 PRIMARY KEY, j INT8 NOT NULL UNIQUE);
>     CREATE TABLE child_pk (i INT8 PRIMARY KEY REFERENCES parent (i));
>     CREATE TABLE child_secondary (i INT8 PRIMARY KEY, j INT8 REFERENCES parent (i));
> ```

Did you intend for this to be `parent (j)`?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60510)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @postamar)

___
*[pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MTImcTLDskL0WuhKTEj:-MTN4u5r1F8b2qEM0YCe:b-hbs7yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ce7be2a8ef42a60f91045cf1e9985785940b7f/pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go#L99)):*
> Arguably there should be a testing library that can do all the work of injecting descriptors/namespace entries into a test server. 

If I implemented that using a big query, would it drive you nuts? 


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60510)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MTImcTLDskL0WuhKTEj:-MTN9jv47EXrxFBegqp0:bdr48dd) ([raw file](https://github.com/cockroachdb/cockroach/blob/49ce7be2a8ef42a60f91045cf1e9985785940b7f/pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go#L99)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> Arguably there should be a testing library that can do all the work of injecting descriptors/namespace entries into a test server. 

If I implemented that using a big query, would it drive you nuts? 
</blockquote></details>

I don't know, I think the point about being able to understand and modify the test easily still stands. I also think it's best to not expose ourselves to potential SQL bugs (like the thing about anti-joins on system tables) in something like this. But it's up to you.


<!-- Sent from Reviewable.io -->

Commit:
sql: fixed a bug in crdb_internal.unsafe_upsert_namespace_entry validation

This bug would have prevented injecting a namespace entry for tables or
types in user-defined schemas.

Release note (bug fix): Fixed a bug in
`crdb_internal.unsafe_upsert_namespace_entry` related to tables and types
in user-defined schemas.

Commit:
testutils/sqlutils: add InjectDescriptors to inject a set of descriptors

This adds a function which takes protobuf descriptors and injects them using
the SQL interface. It resets versions and modifications times as well as
synthesizes public schemas.

Release note: None

ajwerner commented:
I added an injection toolkit in `sqlutils` and uncovered a bug in repair which I'll backport in the first commit. RFAL

thoszhang reviewed (commented):
> Reviewed 1 of 6 files at r2, 6 of 11 files at r4, 1 of 3 files at r6, 10 of 10 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60510)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/kv/kvserver/store.go, line 1364 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MUEbdaKDn-3gfMIRUkv:-MUEbdaKDn-3gfMIRUkw:bbg7k3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/960d50cb40c9814e47e3052297b906920191412a/pkg/kv/kvserver/store.go#L1364)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
> 	value, _, err := storage.MVCCGet(ctx, eng, keys.StoreIdentKey(), hlc.Timestamp{},  storage.MVCCGetOptions{})
> 	found := value != nil
> 	// If we found a result, parse it regardless of the error returned by MVCCGet.
> 	if found {
> 		// If the unmarshal failed, return its result. Otherwise, pass
> 		// through the underlying error (which may be a WriteIntentError
> 		// to be handled specially alongside the returned value).
> 		if err := value.GetProto(&ident); err != nil {
> 			return found, err
> 		}
> 	}
> ```
</details>

What does this do?
___
*[pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go, line 40 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MUEzC311oXbLVBOQ8G-:-MUEzC311oXbLVBOQ8G0:banjls8) ([raw file](https://github.com/cockroachdb/cockroach/blob/72196791bdaeafe0492212b3846744c16ea75b89/pkg/migration/migrations/foreign_key_representation_upgrade_external_test.go#L40)):*
> ```Go
> 	   	BEGIN;
> 	   		CREATE TABLE parent (i INT8 PRIMARY KEY, j INT8 NOT NULL UNIQUE);
> 	   		CREATE TABLE child_pk (i INT8 PRIMARY KEY REFERENCES parent (j));
> ```

Did you intend to also change this line? Previously it was `parent (i)`.
___
*[pkg/migration/migrations/migrations.go, line 58 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MUEzQmd14IqnXAuzjU-:-MUEzQme3N9frYMyfK-X:bbtzs8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/72196791bdaeafe0492212b3846744c16ea75b89/pkg/migration/migrations/migrations.go#L58)):*
> ```Go
> 		"upgrade old foreign key representation",
> 		toCV(clusterversion.ForeignKeyRepresentationMigration),
> 		foreignKeyRepresentationUpgrade),
> ```

nit: newline before the `)` for consistency


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60510)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/kv/kvserver/store.go, line 1364 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60510#-MUEbdaKDn-3gfMIRUkv:-MUF-f89AuBXTewX_n84:b-2nlqqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/960d50cb40c9814e47e3052297b906920191412a/pkg/kv/kvserver/store.go#L1364)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

> ```
> 	value, _, err := storage.MVCCGet(ctx, eng, keys.StoreIdentKey(), hlc.Timestamp{},  storage.MVCCGetOptions{})
> 	found := value != nil
> 	// If we found a result, parse it regardless of the error returned by MVCCGet.
> 	if found {
> 		// If the unmarshal failed, return its result. Otherwise, pass
> 		// through the underlying error (which may be a WriteIntentError
> 		// to be handled specially alongside the returned value).
> 		if err := value.GetProto(&ident); err != nil {
> 			return found, err
> 		}
> 	}
> ```
What does this do?
</blockquote></details>

Never mind, I'm not sure why Reviewable showed me this.


<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r=lucy-zhang

ajwerner commented:
bors r-

ajwerner commented:
bors r+
